src/XR/features/WebXRFeaturePointSystem.ts

```diff
@@ -0,0 +1,176 @@
+     * Will notify the observers about which feature points have been added.
+     */
+    public readonly onFeaturePointsAddedObservable: Observable<number[]> = new Observable();
+    /**
+     * Observers registered here will be executed whenever a feature point has been updated (on XRFrame while the session is tracking).
+     * Will notify the observers about which feature points have been updated.
+     */
+    public readonly onFeaturePointsUpdatedObservable: Observable<number[]> = new Observable();
+    /**
+     * The currrent feature point cloud maintained across frames.
+     */
+    public readonly featurePointCloud: Array<IWebXRFeaturePoint> = this._featurePointCloud;
+
+    /**
+     * construct the feature point system
+     * @param _xrSessionManager an instance of xr Session manager
+     */
+    constructor(_xrSessionManager: WebXRSessionManager) {
+        super(_xrSessionManager);
+        if (this._xrSessionManager.session) {
+            this._init();
+        } else {
+            this._xrSessionManager.onXRSessionInit.addOnce(() => {
+                this._init();
+            });
+        }
+    }
+
+    /**
+     * Detach this feature.
+     * Will usually be called by the features manager
+     *
+     * @returns true if successful.
+     */
+    public detach(): boolean {
+        if (!super.detach()) {
+            return false;
+        }
+
+        this.featurePointCloud.length = 0;
+        return true;
+    }
+
+    /**
+     * Dispose this feature and all of the resources attached
+     */
+    public dispose(): void {
+        super.dispose();
+
+        this._featurePointCloud.length = 0;
+        this.onFeaturePointsUpdatedObservable.clear();
+    }
+
+    /**
+     * On receiving a new XR frame if this feature is attached notify observers new feature point data is available.
+     */
+    protected _onXRFrame(frame: XRFrame) {
+        if (!this.attached || !this._enabled || !frame) {
+            return;
+        }
+
+        const featurePointRawData: number[] | undefined = frame.featurePointCloud;
+        if (!featurePointRawData || featurePointRawData.length == 0) {
+            return;
+        } else {
+            if (featurePointRawData.length % 5 != 0) {
+                throw new Error("Received malformed feature point cloud of length: " + featurePointRawData.length);
+            }
+
+            const numberOfFeaturePoints : number = featurePointRawData.length / 5;
+            let updatedFeaturePoints = new Array();
+            let addedFeaturePoints = new Array();
+            for (var i = 0; i < numberOfFeaturePoints; i++) {
+                const rawIndex: number = i * 5;
+
+                const id = featurePointRawData[rawIndex + 4];
+                // IDs should be durable across frames and strictly increasing from 0 up, so use them as indexing into the feature point array.
+                if (id == this._featurePointCloud.length) {

```

Comment:
You could do something like this and save the ID comparison:

```javascript
// create a new point, setting the ID and 0-values for the rest (will be updated right after)
this._featurePointCloud[id] = this._featurePointCloud[id] || { id, confidenceValue: 0, position: new Vector() };
this._featurePointCloud[id].position.x = featurePointRawData[rawIndex];
this._featurePointCloud[id].position.y = featurePointRawData[rawIndex + 1];
this._featurePointCloud[id].position.z = featurePointRawData[rawIndex + 2];
this._featurePointCloud[id].confidenceValue = featurePointRawData[rawIndex + 3];
updatedFeaturePoints.push(id);
```

Reply:
I need to be able to separate them into added an updated feature points so I still need that first safety check, but I do like this pattern of instantiating it and not having two different sets of code populating the values.